pkg/mls/api/v1/service.go

```diff
@@ -77,220 +157,354 @@ func (s *Service) FetchKeyPackages(ctx context.Context, req *proto.FetchKeyPacka
 	if err != nil {
-		// TODO: Separate validation errors from internal errors
-		return nil, status.Errorf(codes.InvalidArgument, "invalid group message: %s", err)
+		return nil, status.Errorf(codes.Internal, "failed to get identity updates: %s", err)
 	}
 
-	for i, result := range validationResults {
-		message := messages[i]
+	resUpdates := make([]*mlsv1.GetIdentityUpdatesResponse_WalletUpdates, len(accountAddresses))
+	for i, accountAddress := range accountAddresses {
+		walletUpdates := updates[accountAddress]
 
-		if err = requireReadyToSend(result.GroupId, message); err != nil {
-			return nil, err
+		resUpdates[i] = &mlsv1.GetIdentityUpdatesResponse_WalletUpdates{
+			Updates: []*mlsv1.GetIdentityUpdatesResponse_Update{},
 		}
 
-		// TODO: Wrap this in a transaction so publishing is all or nothing
-		if err = s.publishMessage(ctx, topic.BuildGroupTopic(result.GroupId), message); err != nil {
-			return nil, status.Errorf(codes.Internal, "failed to publish message: %s", err)
+		for _, walletUpdate := range walletUpdates {
+			resUpdates[i].Updates = append(resUpdates[i].Updates, buildIdentityUpdate(walletUpdate))
 		}
 	}
 
-	return &emptypb.Empty{}, nil
+	return &mlsv1.GetIdentityUpdatesResponse{
+		Updates: resUpdates,
+	}, nil
 }
 
-func (s *Service) publishMessage(ctx context.Context, contentTopic string, message []byte) error {
-	log := s.log.Named("publish-mls").With(zap.String("content_topic", contentTopic))
-	env, err := s.messageStore.InsertMLSMessage(ctx, contentTopic, message)
-	if err != nil {
-		return status.Errorf(codes.Internal, "failed to insert message: %s", err)
+func (s *Service) SendGroupMessages(ctx context.Context, req *mlsv1.SendGroupMessagesRequest) (res *emptypb.Empty, err error) {
+	if err = validateSendGroupMessagesRequest(req); err != nil {
+		return nil, err
 	}
 
-	if _, err = s.waku.Relay().Publish(ctx, &wakupb.WakuMessage{
-		ContentTopic: contentTopic,
-		Timestamp:    int64(env.TimestampNs),
-		Payload:      message,
-	}); err != nil {
-		return status.Errorf(codes.Internal, "failed to publish message: %s", err)
+	validationResults, err := s.validationService.ValidateGroupMessages(ctx, req.Messages)
+	if err != nil {
+		// TODO: Separate validation errors from internal errors
+		return nil, status.Errorf(codes.InvalidArgument, "invalid group message: %s", err)
 	}
 
-	metrics.EmitPublishedEnvelope(ctx, log, env)
+	for i, result := range validationResults {
+		input := req.Messages[i]
 
-	return nil
-}
+		if err = requireReadyToSend(result.GroupId, input.GetV1().Data); err != nil {
+			return nil, err
+		}
 
-func (s *Service) PublishWelcomes(ctx context.Context, req *proto.PublishWelcomesRequest) (res *emptypb.Empty, err error) {
-	if err = validatePublishWelcomesRequest(req); err != nil {
-		return nil, err
-	}
+		// TODO: Wrap this in a transaction so publishing is all or nothing
+		decodedGroupId, err := hex.DecodeString(result.GroupId)
+		if err != nil {
+			return nil, err
```

Comment:
We can add it in a subsequent PR, but would be nice to have these return GRPC status codes to the client. The raw errors will all just turn into `INTERNAL`